### PR TITLE
[DEVHAS-488]annotation to force generate gitops

### PR DESCRIPTION
**Pull request by yangcao77**
### What does this PR do?:

this PR checks for annotation `forceGitopsGeneration` in component CR to force generate the gitops resources upon reconcilation being hit

verified in log, once the annotation set to true, the re-generation will be triggered
```
{"level":"info","ts":"2023-11-07T18:43:50.703Z","msg":"forceGenerateGitopsResource is true","controller":"component","controllerGroup":"appstudio.redhat.com","controllerKind":"Component","Component":{"name":"devfile-sample-go-basic3","namespace":"application-service-system"},"namespace":"application-service-system","name":"devfile-sample-go-basic3","reconcileID":"897da944-25ef-4df4-b06a-858e5fb8a7e4"}
{"level":"info","ts":"2023-11-07T18:43:50.703Z","logger":"controllers.Application","msg":"API Resource changed: Update","namespace":"application-service-system","audit":"true","name":"devfile-sample-go-basic3","controllerKind":"Application","action":"Update"}
{"level":"info","ts":"2023-11-07T18:43:50.764Z","msg":"Starting reconcile loop for application-service-system/devfile-sample-go-basic3","controller":"component","controllerGroup":"appstudio.redhat.com","controllerKind":"Component","Component":{"name":"devfile-sample-go-basic3","namespace":"application-service-system"},"namespace":"application-service-system","name":"devfile-sample-go-basic3","reconcileID":"897da944-25ef-4df4-b06a-858e5fb8a7e4"}
{"level":"info","ts":"2023-11-07T18:43:50.764Z","msg":"Re-attempting GitOps generation for devfile-sample-go-basic3","controller":"component","controllerGroup":"appstudio.redhat.com","controllerKind":"Component","Component":{"name":"devfile-sample-go-basic3","namespace":"application-service-system"},"namespace":"application-service-system","name":"devfile-sample-go-basic3","reconcileID":"897da944-25ef-4df4-b06a-858e5fb8a7e4"}
{"level":"info","ts":"2023-11-07T18:43:50.768Z","msg":"reading the kubernetes inline from component kubernetes-deploy","controller":"component","controllerGroup":"appstudio.redhat.com","controllerKind":"Component","Component":{"name":"devfile-sample-go-basic3","namespace":"application-service-system"},"namespace":"application-service-system","name":"devfile-sample-go-basic3","reconcileID":"897da944-25ef-4df4-b06a-858e5fb8a7e4"}
{"level":"info","ts":"2023-11-07T18:43:51.164Z","msg":"GitOps re-generation successful for devfile-sample-go-basic3","controller":"component","controllerGroup":"appstudio.redhat.com","controllerKind":"Component","Component":{"name":"devfile-sample-go-basic3","namespace":"application-service-system"},"namespace":"application-service-system","name":"devfile-sample-go-basic3","reconcileID":"897da944-25ef-4df4-b06a-858e5fb8a7e4"}
{"level":"info","ts":"2023-11-07T18:43:51.178Z","msg":"Re-attempting GitOps generation for devfile-sample-go-basic3","controller":"component","controllerGroup":"appstudio.redhat.com","controllerKind":"Component","Component":{"name":"devfile-sample-go-basic3","namespace":"application-service-system"},"namespace":"application-service-system","name":"devfile-sample-go-basic3","reconcileID":"897da944-25ef-4df4-b06a-858e5fb8a7e4"}
{"level":"info","ts":"2023-11-07T18:43:51.184Z","msg":"reading the kubernetes inline from component kubernetes-deploy","controller":"component","controllerGroup":"appstudio.redhat.com","controllerKind":"Component","Component":{"name":"devfile-sample-go-basic3","namespace":"application-service-system"},"namespace":"application-service-system","name":"devfile-sample-go-basic3","reconcileID":"897da944-25ef-4df4-b06a-858e5fb8a7e4"}
{"level":"info","ts":"2023-11-07T18:43:51.518Z","msg":"GitOps re-generation successful for devfile-sample-go-basic3","controller":"component","controllerGroup":"appstudio.redhat.com","controllerKind":"Component","Component":{"name":"devfile-sample-go-basic3","namespace":"application-service-system"},"namespace":"application-service-system","name":"devfile-sample-go-basic3","reconcileID":"897da944-25ef-4df4-b06a-858e5fb8a7e4"}
```

### Which issue(s)/story(ies) does this PR fixes:

https://issues.redhat.com/browse/DEVHAS-488

### PR acceptance criteria:

- [ ] Unit/Functional tests

  

- [ ] Documentation 

   

- [ ] Client Impact

### CODE REVIEW

File: controllers/component_controller.go [hunk 1]
```diff
@@ -754,3 +764,21 @@ func (r *ComponentReconciler) incrementCounterAndRequeue(log logr.Logger, ctx co
 		return ctrl.Result{}, componentErr
 	}
 }
+
+// getForceGenerateGitopsAnnotation gets the internal annotation on the component whether to force generate the gitops resource
+func getForceGenerateGitopsAnnotation(component appstudiov1alpha1.Component) (bool, error) {
```
Comment (johnmcollier) [inline controllers/component_controller.go:769]: We can simplify this a fair bit by avoiding casting the string to bool (which means we also don't need to return an err)

```
func getForceGenerateGitopsAnnotation(component appstudiov1alpha1.Component) bool {
    if compAnnotations == nil && compAnnotations[forceGenerationAnnotation] == "true" {
        return true
    }
    return false
}
```

**Review by johnmcollier — COMMENTED** [review]
Left some comments

File: controllers/component_controller.go [hunk 2]
```diff
@@ -234,6 +243,7 @@ func (r *ComponentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 				return ctrl.Result{}, err
 			} else {
 				log.Info(fmt.Sprintf("GitOps re-generation successful for %s", component.Name))
+				setForceGenerateGitopsAnnotation(&component, "false")
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply controllers/component_controller.go:246]: I'm thinking the person or HAC can enable the force generation, but they are not going to track it? and after the regeneration is done, should this be set to `false`? or the expected usage will be after enabling, the gitops resource going to attempt regeneration in every following reconcile?

[the remark this answers is not in the document]
Comment (johnmcollier) [reply controllers/component_controller.go:246]: > the expected usage will be after enabling, the gitops resource going to attempt regeneration in every following reconcile

I think the expected usage should be that as long as the annotation is present, every reconcile will re-generate the gitops resources. It's up to the user or tool to remove it.

[the remark this answers is not in the document]
Comment (yangcao77) [reply controllers/component_controller.go]: have discussed in team meeting, setting the annotation to false when a force generation is done

File: controllers/component_controller.go (hunk 1 above)
Comment (yangcao77) [reply controllers/component_controller.go]: updated

**Review by johnmcollier — COMMENTED** [review]
Generally looks good to me. Though I think tests need to be added? (Now that the status update works)

**Comment by yangcao77** [comment]
@johnmcollier The status update still not working. I found out any update to the CR causes the reconcile being entered twice. Since we are resetting the annotation back to `false` after the first generation, the second time will overwrite the status to be the one without force generation information. 

I'm not sure why the reconcile is executed twice, verified I see this behavior on main as well. Assuming this is a bug, but should we hold the force annotation until that bug being fixed?  or I can check the timestamp of the gitopsGeneration status in the test to verify the gitops resource has been regenerated

**Comment by yangcao77** [comment]
@johnmcollier just pushed new commits to update the condition type and also controller tests. please help take another look when you have time. thx

**Review by johnmcollier — APPROVED** [review]

**Pull request merged**